Source/DafnyCore/AST/DafnyAst.cs

```diff
@@ -22,26 +24,75 @@ namespace Microsoft.Dafny {
   [System.AttributeUsage(System.AttributeTargets.Field)]
   public class FilledInDuringResolutionAttribute : System.Attribute { }
 
-  public interface INode {
+  public abstract class INode {
 
     /// <summary>
     /// These children should be such that they contain information produced by resolution such as inferred types
     /// and resolved references. However, they should not be so transformed that source location from the initial
     /// program is lost. As an example, the pattern matching compilation may deduplicate nodes from the original AST,
     /// losing source location information, so those transformed nodes should not be returned by this property.
     /// </summary>
-    IEnumerable<INode> Children { get; }
+    public abstract IEnumerable<INode> Children { get; }
+
+    public IToken? StartToken { get; set; } = Token.NoToken;
+    public IToken? EndToken { get; set; } = Token.NoToken;
+
+    protected IEnumerable<IToken>? ownedTokens;
 
     /// <summary>
-    /// These tokens are "owned" by the node, roughly meaning
-    /// they are not "owned" by any other node normally (typically children)
-    /// Owned tokens will help the formatter determine formatting cues
-    /// when considering that node.
+    /// A token is owned by a node if it was used to parse this node,
+    /// but is not owned by any of this Node's children
     /// </summary>
-    IEnumerable<IToken> OwnedTokens { get; }
+    IEnumerable<IToken> OwnedTokens {
+      get {
+        if (ownedTokens == null) {
+          var startNotOwnedToken = Children.Select(child => child.StartToken).ToImmutableHashSet();
+          var endNotOwnedToken = Children.Select(child => child.EndToken).ToImmutableHashSet();
+
+          var tmpTokens = new List<IToken>();
+          if (StartToken == null) {
+            if (EndToken != null) {
+              if (!endNotOwnedToken.Contains(EndToken)) {
+                tmpTokens.Add(EndToken);
+              }
+            }
+          } else {
+            var canOwnTokens = true;
+            if (!startNotOwnedToken.Contains(StartToken)) {
+              tmpTokens.Add(StartToken);
+            } else {
+              canOwnTokens = false;
+            }
+
+            if (EndToken != null) {
+              var tmpToken = StartToken.Next;
+              while (tmpToken != null && tmpToken != EndToken) {
```

Comment:
> Does this implementation have a N^2 complexity for computing the owned tokens of the entire AST?

Given that fetching from a HashSet is O(1), assuming a bounded number of children, and that children are disjoint, this computation of this per node is in constant time, so it's O(N) complexity for the entire tree.

A big difference is that my implementation is not sensitive to the order of children, compared to yours. So I prefer to keep mine, rather than trying to manually ensure that all the Children properties of nodes describe the children in the visual order.

Reply:
But doesn't the implementation here traverse _all_ the tokens for the top node? Then if the AST is a linked list  wouldn't you traverse the tokens N*N/2 times?

Reply:
> But doesn't the implementation here traverse all the tokens for the top node? Then if the AST is a linked list with each node owning one token, wouldn't you traverse the tokens N*N/2 times when computing all the owned tokens?

Right, I hadn't taken that traversal into account. I fixed my implementation to keep track of the pairs stored as a dictionary, so that I don't need to depend on the order of children, just skip the token ranges they represent.